pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,246 @@ package adapter
+	// transformerURI is the URI for the transformer.
+	// Used for channels.
+	transformerURI string
 
 	// extensions is the converted ExtensionsBased64 value.
 	extensions map[string]string
 
-	// SendMode describes how the adapter sends events.
-	// One of [binary, structured, push]. Default: binary
-	SendMode converters.ModeType `envconfig:"SEND_MODE" default:"binary" required:"true"`
+	// converterType use to select which converter to use.
+	converterType converters.ConverterType
 
-	// MetricsConfigJson is a json string of metrics.ExporterOptions.
-	// This is used to configure the metrics exporter options, the config is
-	// stored in a config map inside the controllers namespace and copied here.
-	MetricsConfigJson string `envconfig:"K_METRICS_CONFIG" required:"true"`
+	// reporter reports metrics to the configured backend.
+	reporter StatsReporter
 
-	// LoggingConfigJson is a json string of logging.Config.
-	// This is used to configure the logging config, the config is stored in
-	// a config map inside the controllers namespace and copied here.
-	LoggingConfigJson string `envconfig:"K_LOGGING_CONFIG" required:"true"`
+	// converter used to convert pubsub messages to CE.
+	converter converters.Converter
 
-	// TracingConfigJson is a JSON string of tracing.Config. This is used to configure tracing. The
-	// original config is stored in a ConfigMap inside the controller's namespace. Its value is
-	// copied here as a JSON string.
-	TracingConfigJson string `envconfig:"K_TRACING_CONFIG" required:"true"`
+	// cancel is function to stop pulling messages.
+	cancel context.CancelFunc
 
-	// Environment variable containing the namespace.
-	Namespace string `envconfig:"NAMESPACE" required:"true"`
+	logger *zap.Logger
 
-	// Environment variable containing the name.
-	Name string `envconfig:"NAME" required:"true"`
+	// name of the resource this receive adapter belong to.
+	namespacedResourceName types.NamespacedName
+}
 
-	// Environment variable containing the resource group. E.g., storages.events.cloud.google.com.
-	ResourceGroup string `envconfig:"RESOURCE_GROUP" default:"pullsubscriptions.pubsub.cloud.google.com" required:"true"`
+// NewAdapter creates a new adapter.
+func NewAdapter(
+	ctx context.Context,
+	subscription *pubsub.Subscription,
+	outbound *nethttp.Client,
+	converter converters.Converter,
+	reporter StatsReporter,
+	namespace Namespace,
+	name Name,
+	sinkURI SinkURI,
+	transformerURI TransformerURI,
+	converterType converters.ConverterType,
+	extensions map[string]string) *Adapter {
+	return &Adapter{
+		subscription:           subscription,
+		outbound:               outbound,
+		converter:              converter,
+		reporter:               reporter,
+		sinkURI:                string(sinkURI),
+		transformerURI:         string(transformerURI),
+		namespacedResourceName: types.NamespacedName{Name: string(name), Namespace: string(namespace)},
+		converterType:          converterType,
+		extensions:             extensions,
```

Comment:
It might simplify the constructor to bundle these into a config struct.

Reply:
Will do an AdapterArgs one and pass it here...

Reply:
done